doc/nanocourse/bitvectors.md

```diff
@@ -141,93 +181,97 @@ In order to support rank and select queries we need two helper data structures (
-and should be preferred to larger typed, to reduces the space of the block array.
+and should be preferred to larger types, to reduce the space of the block data structure.
 
 Since we do not know how large our bitvector might be, one should choose a large data type, like `uint64_t`,
 to store to prefix sums for the superblock values.
 
 \endsolution
 
-Now that we now which data types to work with, let's implement the data structures.
+Now that we know which data types to work with, let's implement the data structures.
 
 \assignment{Compute the block and superblock arrays}
 
-Given a bitvector `B` from the previous assignments that can store `6400 bits` (what's the length?),
-compute the entries of the arrays by iterating over the bitvector.
+Given a bitvector `B` from the previous assignments:
+1. Add the member variables `std::vector<uint16_t> blocks` and `std::vector<uint64_t> superblocks`.
+2. Add a member function `void construct(size_t const block_size = 64u, size_t const superblock_size = 512u)`
+that fills the member variables `blocks` and `superblocks`.
 
 With the following main function:
 
-```cpp
-int main()
-{
-    std::vector<uint64_t> B(6400 / 64, 2);
-
-    std::vector<uint16_t> blocks(6400 / 64, 0);
-    std::vector<uint64_t> superblocks(6400 / 1600, 0);
-
-    construct(blocks, superblocks, B);
-
-    for (auto a : superblocks)
-        std::cout << a << " ";
-    std::cout << std::endl;
-}
-```
+\snippet test/snippet/nanocourse/solution4.cpp main
 
 your program should print the following:
 
 ```
+Superblocks:
 0 25 50 75
-```
 
-If you print your blocks too, there should be 100 entries: `0,1,2,...,24,0,1,2,...24,0,1,2,...,24,0,1,2,...24`.
+Blocks:
+0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
+0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
+0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
+0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
+```
 
-If you are not so experienced with C++, you can use the following backbone:
+If you aren't experienced with C++, you can use the following backbone:
 
 \hint
 
 ```cpp
-void construct(std::vector<uint16_t> & blocks,
-               std::vector<uint64_t> & superblocks,
-               std::vector<uint64_t> const & B)
-{
-    size_t block_pos{0u};
-    size_t superblock_pos{0u};
+    // The member function takes two parameters:
+    // `block_size` which is defaulted to be 64 and `superblock_size` which is defaulted to 512.
+    // Defaulted means that you can also call the function without parameters `construct()`.
+    void construct(size_t const block_size = 64, size_t const superblock_size = 512)
+    {
+        size_t number_of_bits = data.size() * 64;
 
-    uint16_t block_count{0u};
-    uint64_t superblock_count{0u};
+        // resize the two arrays of blocks and superblocks
+        blocks = std::vector<uint16_t>(/*TODO: how many blocks do I need?*/, 0);
```

Comment:
Actually we could just do `blocks.resize(/*TODO: how many blocks do I need?*/)`? 
Same in the solutions.

Reply:
:woman_facepalming:  yes

Reply:
I will also do this in the solution 1 constructor instead of the direct data member initialisation to keep it simple and consistent.